src/piton/jax.py

```diff
@@ -0,0 +1,1048 @@
+import numpy as np
+from jax import core, custom_vjp, debug, grad, lax, nn
+from jax import numpy as jnp
+from jax import value_and_grad, vmap, xla_computation
+from jax.core import ConcreteArray, ShapedArray
+from jax.interpreters import ad, batching, xla
+from jax.lib import xla_client
+
+# Per the jax documentation, we currently don't have good typing for arrays
+Array = Any
+
+from piton.extension.jax import (
+    get_kernels,
+    get_local_attention_data,
+    get_local_attention_shape,
+)
+
+# Globally register all of our custom operators
+for name, value, platform in get_kernels():
+    jax.lib.xla_client.register_custom_call_target(
+        name, value, platform=platform
+    )
+
+
+@partial(
+    jax.custom_vjp,
+    nondiff_argnums=(2,),
+)
+def gather_scatter(a: Array, indices: Array, output_dim: int) -> Array:
+    """Perform a combined gather / scatter add operation.
+
+    Indices is an array of shape (N, 2) for N indices.
+    This operator reads from the index specified in indices[:, 0] and writes them to indices[:, 1].
+
+    output_dim is the size of the ouptupt array.
+
+    Note that this operator requires that indices must be sorted in the order of the write indices (indices[:, 1])
+    """
+    return gather_scatter_fwd(a, indices, output_dim)[0]
+
+
+def gather_scatter_fallback(a: Array, indices: Array, output_dim: int) -> Array:
+    """The python fallback implementation of gather scatter."""
+    a_indices = indices[:, 0]
+    b_indices = indices[:, 1]
+
+    a_vals = a.at[a_indices, :].get(mode="fill", fill_value=0)
+
+    result = jnp.zeros((output_dim, a.shape[1]), dtype=a.dtype)
+
+    return result.at[b_indices, :].add(a_vals, mode="drop")
+
+
+def gather_scatter_fwd(
+    a: Array, indices: Array, output_dim: int
+) -> Tuple[Array, Tuple[int, Array]]:
+    """The forward pass for gather / scatter"""
+    r = gather_scatter_p.bind(a, indices, output_dim=output_dim)
+    return r, (a.shape[0], indices)
+
+
+def gather_scatter_bwd(
+    _output_dim: int, res: Tuple[int, Array], g: Array
+) -> Tuple[Array, None]:
+    """The backward pass for gather / scatter
+
+    The basic idea here is that we transpose the indices and apply it in reverse to get the backward pass.
+    """
+
+    a_dim, indices = res
+
+    indices_a = indices[:, 0]
+    indices_b = indices[:, 1]
+
+    sort_indices = jnp.argsort(indices_a)
+
+    # Note that these are transposed
+    trans_indices = jnp.stack((indices_b, indices_a), axis=-1)
+    trans_indices = trans_indices[sort_indices]
+
+    r = gather_scatter_p.bind(g, trans_indices, output_dim=a_dim)
+
+    return (r, None)
+
+
+gather_scatter.defvjp(gather_scatter_fwd, gather_scatter_bwd)
+
+# The actual primitive used to implement this operation
+gather_scatter_p = jax.core.Primitive("gather_scatter_p")
+
+
+def gather_scatter_abstract_eval(
+    a: Array, indices: Array, output_dim: int
+) -> jax.core.ShapedArray:
+    """The abstract shape computation for gather_scatter."""
```

Comment:
Would add that the shape returned here must match the shape of the return value from `gather_scatter()`

Reply:
That isn't correct. I think this name was misleading. Renamed to gather_scatter_p_abstract_eval. The primitives != the operation itself.